### ISSUE TITLE
Bad Cluster URL Fails To Load All Kubernetes Clusters Resulting in Almost Blank Page

### ISSUE THREAD

**Issue by asaf-erlich** [backstage/backstage#13794]
## Expected Behavior
When clusters are configured in backstage, but the URL is incorrect or for a cluster that has been destroyed so the url is pointing to an incorrect address, I would expect those clusters to not appear on the page but other clusters to appear just fine.

## Actual Behavior
If someone deletes a cluster in aws, something that the maintainers of their backstage instance have no control over, it just starts throwing 500s and all kubernetes tab pages fail to load any clusters. The error it says at the top and in the logs is `Error: getaddrinfo ENOTFOUND badurl.gr7.us-west-2.eks.amazonaws.com`. I realize our use case is unique, but we have many clusters owned by many different teams. It is rare, but happens that a cluster in a non-production account may need to be replaced. After that the old url no longer works. All the other cluster's configuration is correct and I would wish that the bad cluster be simply skipped. I would want all other clusters's resources to be queried successfully and displayed correctly. It would be nice to correctly handle this error gracefully. 

Screenshot:
![Screen Shot 2022-09-21 at 3 09 19 PM](https://user-images.githubusercontent.com/1738335/191591253-b4664744-c1d8-4750-8bde-70ba82c8411e.png)

## Steps to Reproduce
Screenshot above shows an example with a bad url cluster. Steps to somewhat reproduce below, will do my best to include configuration as well.

1. Install [k8s](https://backstage.io/docs/features/kubernetes/overview) backend and frontend plugin following [installation instructions](https://backstage.io/docs/features/kubernetes/installation).
1. [Configure](https://backstage.io/docs/features/kubernetes/configuration) clusters with a similar setup to the one below. Below is the kubernetes part of the app-config.yaml:
    ```yaml
    kubernetes:
      serviceLocatorMethod:
        type: 'multiTenant'
      clusterLocatorMethods:
        - type: 'config'
          clusters:
            - name: clustername
              url: https://badurl.amazonaws.com
              authProvider: 'aws'
              caData: ${CLUSTER_K8S_CONFIG_CA_DATA}
    ```
1. Configure a component to have the appropriate annotations for the k8s plugin. Example catalog-info.yaml file:
    ```yaml
    apiVersion: backstage.io/v1alpha1
    kind: Component
    metadata:
      name: backstage
      description: The Backstage
      annotations:
        backstage.io/kubernetes-label-selector: 'app=backstage'
        backstage.io/kubernetes-namespace: backstage
    spec:
      type: website
      owner: owner
      lifecycle: experimental
    ```

## Context
This is a badly handled error. In the screenshot above only one cluster has a bad url while the others should work fine. Let me know if more information is needed.

## Your Environment
- Browser Information: Google Chrome Version 105.0.5195.102 (Official Build) (arm64)
- Output of `yarn backstage-cli info`: 

```bash
yarn backstage-cli info
yarn run v1.22.17
$ /Users/asaferlich/dev/src/github.com/segmentio/backstage-segment/node_modules/.bin/backstage-cli info
OS:   Darwin 21.6.0 - darwin/arm64
node: v16.14.2
yarn: 1.22.17
cli:  0.18.1 (installed)
backstage:  1.5.1

Dependencies:
  @backstage/app-defaults                          1.0.5
  @backstage/backend-common                        0.15.0
  @backstage/backend-plugin-api                    0.1.1
  @backstage/backend-tasks                         0.3.4
  @backstage/catalog-client                        1.0.4
  @backstage/catalog-model                         1.1.0
  @backstage/cli-common                            0.1.9
  @backstage/cli                                   0.18.1
  @backstage/config-loader                         1.1.3
  @backstage/config                                1.0.1
  @backstage/core-app-api                          1.0.5
  @backstage/core-components                       0.11.0
  @backstage/core-plugin-api                       1.0.5
  @backstage/dev-utils                             1.0.5
  @backstage/errors                                1.1.0
  @backstage/integration-react                     1.1.3
  @backstage/integration                           1.3.0
  @backstage/plugin-analytics-module-ga            0.1.19
  @backstage/plugin-api-docs                       0.8.8
  @backstage/plugin-app-backend                    0.3.35
  @backstage/plugin-auth-backend                   0.15.1
  @backstage/plugin-auth-node                      0.2.4
  @backstage/plugin-catalog-backend-module-github  0.1.6
  @backstage/plugin-catalog-backend                1.3.1
  @backstage/plugin-catalog-common                 1.0.5
  @backstage/plugin-catalog-graph                  0.2.20
  @backstage/plugin-catalog-node                   1.0.1
  @backstage/plugin-catalog-react                  1.1.3
  @backstage/plugin-catalog                        1.5.0
  @backstage/plugin-github-actions                 0.5.8
  @backstage/plugin-github-pull-requests-board     0.1.2
  @backstage/plugin-home                           0.4.24
  @backstage/plugin-kubernetes-backend             0.7.1
  @backstage/plugin-kubernetes-common              0.4.1
  @backstage/plugin-kubernetes                     0.7.1
  @backstage/plugin-org                            0.5.8
  @backstage/plugin-pagerduty                      0.5.1
  @backstage/plugin-permission-common              0.6.3
  @backstage/plugin-permission-node                0.6.4
  @backstage/plugin-permission-react               0.4.4
  @backstage/plugin-proxy-backend                  0.2.29
  @backstage/plugin-scaffolder-backend             1.5.1
  @backstage/plugin-scaffolder-common              1.1.2
  @backstage/plugin-scaffolder                     1.5.0
  @backstage/plugin-search-backend-module-pg       0.3.6
  @backstage/plugin-search-backend-node            1.0.1
  @backstage/plugin-search-backend                 1.0.1
  @backstage/plugin-search-common                  1.0.0
  @backstage/plugin-search-react                   1.0.1
  @backstage/plugin-search                         1.0.1
  @backstage/plugin-shortcuts                      0.3.0
  @backstage/plugin-stack-overflow                 0.1.4
  @backstage/plugin-techdocs-backend               1.2.1
  @backstage/plugin-techdocs-module-addons-contrib 1.0.3
  @backstage/plugin-techdocs-node                  1.3.0
  @backstage/plugin-techdocs-react                 1.0.3
  @backstage/plugin-techdocs                       1.3.1
  @backstage/plugin-user-settings                  0.4.7
  @backstage/release-manifests                     0.0.5
  @backstage/test-utils                            1.1.3
  @backstage/theme                                 0.2.16
  @backstage/types                                 1.0.0
  @backstage/version-bridge                        1.0.1
✨  Done in 0.96s.
```

Note: This issue should probably have the `k8s-plugin` label but I don't think I can add it myself.

**Comment by freben** [comment]
ping @backstage/warpspeed

**Comment by freben** [comment]
@asaf-erlich This sounds like a very worthwhile fix. Would you like to look into making a PR for it?

There are a few `Promise.all` that might apply, like [this one](https://github.com/backstage/backstage/blob/aa5679b64fa6e98fdfb2abba080eabb0343bc9f4/plugins/kubernetes-backend/src/service/KubernetesFanOutHandler.ts#L250), or [this one](https://github.com/backstage/backstage/blob/aa5679b64fa6e98fdfb2abba080eabb0343bc9f4/plugins/kubernetes-backend/src/service/KubernetesFetcher.ts#L104).

**Comment by asaf-erlich** [comment]
Hello @freben , my team has a few quarterly goals we are trying to wrap up, but I plan to work on issues like this one in October, which is only a few weeks away. Thanks for sharing those two locations that will help me track down where the fix is probably located.

**Comment by mclarke47** [comment]
I think a fix here would be great, like you say these errors should be bundled in the error section of the cluster response, they shouldn't be a show stopping error

**Comment by namco1992** [comment]
Seems to be related to #13794 too.

**Comment by jamieklassen** [comment]
@namco1992 did you mean to post that comment somewhere else, or link to a different issue? This is #13794 right here!

**Comment by asaf-erlich** [comment]
I believe he meant https://github.com/backstage/backstage/issues/13767 which mentions this issue above.

**Comment by jamieklassen** [comment]
I have a fix for this on [this branch](https://github.com/jamieklassen/backstage/commits/handle-k8s-transport-errors) of my work. I'm not happy with it yet, but y'all can assign it to me as I expect to have a professional implementation by the end of the week.

**Comment by svyotov** [comment]
We also have a very similar issue, I have a PR coming, where we check if a namespace exists. If the namespace does not exist, then there is no need to bombard the user with exceptions about non existing resources query.

**Comment by jhaals** [comment]
Is this fixed by #13966?

**Comment by asaf-erlich** [comment]
Maybe, would need to test it. Will that fix be in the next release or one that is already out?

**Comment by jamieklassen** [comment]
No, this is not fixed by #13966, but it's fixed by #14451!

**Comment by asaf-erlich** [comment]
Just tested this locally and confirm this fixed the issue. Updated screenshot below.
![Screenshot 2023-01-03 at 3 49 04 PM](https://user-images.githubusercontent.com/1738335/210440327-9eb16dbe-c9f3-4a4f-b9bf-87f6368d06a3.png)

So it defaults to the same behavior as https://github.com/backstage/backstage/issues/13792 where the cluster appears with 0 resources. This is a nice step in the right direction. Thank you everyone who submitted fixes so far that contributed to this improved behavior.
